File: Quick/QuickSortStep.java
import java.io.*;
import java.util.*;

public class QuickSortStep {

    public static List<String> loadDataset(String filename, int startRow, int endRow) {
        List<String> data = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            int idx = 1;
            while ((line = br.readLine()) != null) {
                if (idx >= startRow && idx <= endRow) {
                    String[] parts = line.split(",");
                    data.add(parts[0] + "/" + parts[1]);
                }
                if (idx > endRow) break;
                idx++;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return data;
    }

    public static void recordSteps(List<String> steps, String filename) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (String step : steps) {
                writer.write(step);
                writer.newLine();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static long extractNumber(String element) {
        return Long.parseLong(element.split("/")[0]);
    }

    public static int partition(List<String> arr, int low, int high) {
        long pivot = extractNumber(arr.get(high)); // change to long
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (extractNumber(arr.get(j)) <= pivot) {
                i++;
                Collections.swap(arr, i, j);
            }
        }

        Collections.swap(arr, i + 1, high);
        return i + 1;
    }

    public static void quickSortWithSteps(List<String> arr, int low, int high, List<String> steps) {
        if (low < high) {
            int pi = partition(arr, low, high);
            steps.add("pi=" + (pi + 1) + " [" + String.join(", ", arr) + "]");
            quickSortWithSteps(arr, low, pi - 1, steps);
            quickSortWithSteps(arr, pi + 1, high, steps);
        }
    }

    public static void main(String[] args) {
        String datasetFilename = "dataset/generate_dataset/dataset_sample_1000.csv";
        int startRow = 1;
        int endRow = 7;
        String outputFilename = "dataset/quick_sort_result/quick_sort_step_" + startRow + "_" + endRow + ".txt";

        List<String> data = loadDataset(datasetFilename, startRow, endRow);
        List<String> steps = new ArrayList<>();
        steps.add("[" + String.join(", ", data) + "]");

        quickSortWithSteps(data, 0, data.size() - 1, steps);
        recordSteps(steps, outputFilename);
    }
}

File: Quick/quick_sort_step.py
import csv

def load_dataset(filename, start_row, end_row):
    data = []
    with open(filename, mode='r') as file:
        reader = csv.reader(file)
        for idx, row in enumerate(reader, start=1):
            if start_row <= idx <= end_row:
                data.append(f"{row[0]}/{row[1]}")
            if idx > end_row:
                break
    return data

def extract_number(element):
    return int(element.split('/')[0])

def record_steps(steps, filename):
    with open(filename, 'w') as file:
        for step in steps:
            if isinstance(step, str):
                file.write(step + '\n')
            else:
                formatted_step = '[' + ', '.join(step) + ']\n'
                file.write(formatted_step)

def quick_sort_with_steps(arr, low, high, steps):
    if low < high:
        pi = partition(arr, low, high)
        steps.append(f"pi={pi} [{', '.join(arr)}]")
        quick_sort_with_steps(arr, low, pi - 1, steps)
        quick_sort_with_steps(arr, pi + 1, high, steps)


def partition(arr, low, high):
    pivot = extract_number(arr[high])
    i = low - 1

    for j in range(low, high):
        if extract_number(arr[j]) <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

dataset_filename = "dataset/generate_dataset/dataset_sample_1000.csv"
start_row = 1
end_row = 7
output_filename = f"dataset/quick_sort_result/quick_sort_step_{start_row}_{end_row}.txt"

data = load_dataset(dataset_filename, start_row, end_row)
steps = []
steps.append(data.copy())

quick_sort_with_steps(data, 0, len(data) - 1, steps)
record_steps(steps, output_filename)
